src/particles.cpp

```diff
@@ -431,50 +336,245 @@ void ParticleSpawner::step(float dtime, ClientEnvironment &env)
+			i->second->step(dtime, m_env);
 			i++;
 		}
 	}
 }
 
-void delete_particlespawner (u32 id)
+void ParticleManager::stepParticles (float dtime)
 {
-	if (all_particlespawners.find(id) != all_particlespawners.end())
+	JMutexAutoLock lock(m_particlelistlock);
+	for(std::vector<Particle*>::iterator i = m_particles.begin();
+			i != m_particles.end();)
 	{
-		delete all_particlespawners.find(id)->second;
-		all_particlespawners.erase(id);
+		if ((*i)->get_expired())
+		{
+			(*i)->remove();
+			delete *i;
+			i = m_particles.erase(i);
+		}
+		else
+		{
+			(*i)->step(dtime);
+			i++;
+		}
 	}
 }
 
-void clear_particles ()
+void ParticleManager::clearAll ()
 {
+	JMutexAutoLock lock(m_spawnerlistlock);
+	JMutexAutoLock lock2(m_particlelistlock);
 	for(std::map<u32, ParticleSpawner*>::iterator i =
-			all_particlespawners.begin();
-			i != all_particlespawners.end();)
+			m_particlespawners.begin();
+			i != m_particlespawners.end();)
 	{
 		delete i->second;
-		all_particlespawners.erase(i++);
+		m_particlespawners.erase(i++);
 	}
 
 	for(std::vector<Particle*>::iterator i =
-			all_particles.begin();
-			i != all_particles.end();)
+			m_particles.begin();
+			i != m_particles.end();)
 	{
 		(*i)->remove();
 		delete *i;
-		i = all_particles.erase(i);
+		i = m_particles.erase(i);
+	}
+}
+
+void ParticleManager::handleParticleEvent(ClientEvent *event, IGameDef *gamedef,
+		scene::ISceneManager* smgr, LocalPlayer *player)
+{
+	if (event->type == CE_DELETE_PARTICLESPAWNER) {
+		JMutexAutoLock lock(m_spawnerlistlock);
+		if (m_particlespawners.find(event->delete_particlespawner.id) !=
+				m_particlespawners.end())
+		{
+			delete m_particlespawners.find(event->delete_particlespawner.id)->second;
+			m_particlespawners.erase(event->delete_particlespawner.id);
+		}
+		// no allocated memory in delete event
+		return;
 	}
+
+	if (event->type == CE_ADD_PARTICLESPAWNER) {
+
+		{
+			JMutexAutoLock lock(m_spawnerlistlock);
+			if (m_particlespawners.find(event->delete_particlespawner.id) !=
+							m_particlespawners.end())
+			{
+				delete m_particlespawners.find(event->delete_particlespawner.id)->second;
+				m_particlespawners.erase(event->delete_particlespawner.id);
+			}
+		}
+		video::ITexture *texture =
+			gamedef->tsrc()->getTexture(*(event->add_particlespawner.texture));
+
+		ParticleSpawner* toadd = new ParticleSpawner(gamedef, smgr, player,
+				event->add_particlespawner.amount,
+				event->add_particlespawner.spawntime,
+				*event->add_particlespawner.minpos,
+				*event->add_particlespawner.maxpos,
+				*event->add_particlespawner.minvel,
+				*event->add_particlespawner.maxvel,
+				*event->add_particlespawner.minacc,
+				*event->add_particlespawner.maxacc,
+				event->add_particlespawner.minexptime,
+				event->add_particlespawner.maxexptime,
+				event->add_particlespawner.minsize,
+				event->add_particlespawner.maxsize,
+				event->add_particlespawner.collisiondetection,
+				event->add_particlespawner.vertical,
```

Comment:
Hmm thinking about once again I don't think it'd be a good idee to change it. In order to do this I'd have to define a new struct because I can't use the add_particle_spawner struct as it's part of the ClientEvent union.